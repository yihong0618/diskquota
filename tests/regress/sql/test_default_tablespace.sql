-- test role_tablespace_quota works with tables/databases in default tablespace
-- test role_tablespace_quota works with tables/databases in non-default tablespace with hard limits on

-- start_ignore
\! mkdir -p /tmp/custom_tablespace
-- end_ignore

DROP ROLE if EXISTS role1;
DROP ROLE if EXISTS role2;
CREATE ROLE role1 SUPERUSER;
CREATE ROLE role2 SUPERUSER;
SET ROLE role1;

DROP TABLE if EXISTS t;
CREATE TABLE t (i int) DISTRIBUTED BY (i);

-- with hard limits off
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null

SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to success
INSERT INTO t SELECT generate_series(1, 100);
INSERT INTO t SELECT generate_series(1, 1000000);
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);

SELECT r.rolname, t.spcname, b.target_type
FROM diskquota.rejectmap AS b, pg_tablespace AS t, pg_roles AS r
WHERE b.tablespace_oid = t.oid AND b.target_oid = r.oid AND r.rolname = 'role1'
ORDER BY r.rolname, t.spcname, b.target_type;

DROP TABLE IF EXISTS t;
SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '-1');

SET ROLE role2;
CREATE TABLE t (i int) DISTRIBUTED BY (i);

-- with hard limits on
\! gpconfig -c "diskquota.hard_limit" -v "on" > /dev/null
\! gpstop -u > /dev/null

SELECT diskquota.set_role_tablespace_quota('role2', 'pg_default', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to fail because of hard limits
INSERT INTO t SELECT generate_series(1, 50000000);
DROP TABLE IF EXISTS t;

SET ROLE role1;
-- database in customized tablespace
CREATE TABLESPACE custom_tablespace LOCATION '/tmp/custom_tablespace';
CREATE DATABASE db_with_tablespace TABLESPACE custom_tablespace;
\c db_with_tablespace;
SET ROLE role1;
CREATE EXTENSION diskquota;

-- with hard limits off
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null

SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to success
CREATE TABLE t_in_custom_tablespace (i) AS SELECT generate_series(1, 100) DISTRIBUTED BY (i);
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to fail
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);

SELECT r.rolname, t.spcname, b.target_type
FROM diskquota.rejectmap AS b, pg_tablespace AS t, pg_roles AS r
WHERE b.tablespace_oid = t.oid AND b.target_oid = r.oid AND r.rolname = 'role1'
ORDER BY r.rolname, t.spcname, b.target_type;

DROP TABLE IF EXISTS t_in_custom_tablespace;
SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '-1');
SELECT diskquota.wait_for_worker_new_epoch();
SET ROLE role2;

-- with hard limits on
\! gpconfig -c "diskquota.hard_limit" -v "on" > /dev/null
\! gpstop -u > /dev/null

SELECT diskquota.set_role_tablespace_quota('role2', 'custom_tablespace', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();

DROP TABLE IF EXISTS t_in_custom_tablespace;
-- expect insert to fail because of hard limits
CREATE TABLE t_in_custom_tablespace (i) AS SELECT generate_series(1, 50000000) DISTRIBUTED BY (i);

-- clean up
DROP TABLE IF EXISTS t_in_custom_tablespace;

\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null

SELECT diskquota.pause();
SELECT diskquota.wait_for_worker_new_epoch();
DROP EXTENSION IF EXISTS diskquota;

\c contrib_regression;
SELECT diskquota.wait_for_worker_new_epoch();
DROP DATABASE IF EXISTS db_with_tablespace;
DROP TABLESPACE IF EXISTS custom_tablespace;

RESET ROLE;
DROP ROLE IF EXISTS role1;
DROP ROLE IF EXISTS role2;
