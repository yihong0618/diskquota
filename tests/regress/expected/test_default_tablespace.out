-- test role_tablespace_quota works with tables/databases in default tablespace
-- test role_tablespace_quota works with tables/databases in non-default tablespace with hard limits on
-- start_ignore
\! mkdir -p /tmp/custom_tablespace
-- end_ignore
DROP ROLE if EXISTS role1;
NOTICE:  role "role1" does not exist, skipping
DROP ROLE if EXISTS role2;
NOTICE:  role "role2" does not exist, skipping
CREATE ROLE role1 SUPERUSER;
CREATE ROLE role2 SUPERUSER;
SET ROLE role1;
DROP TABLE if EXISTS t;
NOTICE:  table "t" does not exist, skipping
CREATE TABLE t (i int) DISTRIBUTED BY (i);
-- with hard limits off
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null
SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to success
INSERT INTO t SELECT generate_series(1, 100);
INSERT INTO t SELECT generate_series(1, 1000000);
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);
ERROR:  tablespace:pg_default role:role1 diskquota exceeded
SELECT r.rolname, t.spcname, b.target_type
FROM diskquota.blackmap AS b, pg_tablespace AS t, pg_roles AS r
WHERE b.tablespace_oid = t.oid AND b.target_oid = r.oid AND r.rolname = 'role1'
ORDER BY r.rolname, t.spcname, b.target_type;
 rolname |  spcname   |      target_type      
---------+------------+-----------------------
 role1   | pg_default | ROLE_TABLESPACE_QUOTA
(1 row)

DROP TABLE IF EXISTS t;
SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '-1');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SET ROLE role2;
CREATE TABLE t (i int) DISTRIBUTED BY (i);
-- with hard limits on
\! gpconfig -c "diskquota.hard_limit" -v "on" > /dev/null
\! gpstop -u > /dev/null
SELECT diskquota.set_role_tablespace_quota('role2', 'pg_default', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to fail because of hard limits
INSERT INTO t SELECT generate_series(1, 50000000);
ERROR:  tablespace:1663 role:3143588 diskquota exceeded  (seg0 127.0.0.1:6002 pid=2298)
DROP TABLE IF EXISTS t;
SET ROLE role1;
-- database in customized tablespace
CREATE TABLESPACE custom_tablespace LOCATION '/tmp/custom_tablespace';
CREATE DATABASE db_with_tablespace TABLESPACE custom_tablespace;
\c db_with_tablespace;
SET ROLE role1;
CREATE EXTENSION diskquota;
-- with hard limits off
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null
SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to success
CREATE TABLE t_in_custom_tablespace (i) AS SELECT generate_series(1, 100) DISTRIBUTED BY (i);
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to fail
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
ERROR:  tablespace:custom_tablespace role:role1 diskquota exceeded
SELECT r.rolname, t.spcname, b.target_type
FROM diskquota.blackmap AS b, pg_tablespace AS t, pg_roles AS r
WHERE b.tablespace_oid = t.oid AND b.target_oid = r.oid AND r.rolname = 'role1'
ORDER BY r.rolname, t.spcname, b.target_type;
 rolname |      spcname      |      target_type      
---------+-------------------+-----------------------
 role1   | custom_tablespace | ROLE_TABLESPACE_QUOTA
(1 row)

DROP TABLE IF EXISTS t_in_custom_tablespace;
SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '-1');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

SET ROLE role2;
-- with hard limits on
\! gpconfig -c "diskquota.hard_limit" -v "on" > /dev/null
\! gpstop -u > /dev/null
SELECT diskquota.set_role_tablespace_quota('role2', 'custom_tablespace', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

DROP TABLE IF EXISTS t_in_custom_tablespace;
NOTICE:  table "t_in_custom_tablespace" does not exist, skipping
-- expect insert to fail because of hard limits
CREATE TABLE t_in_custom_tablespace (i) AS SELECT generate_series(1, 50000000) DISTRIBUTED BY (i);
ERROR:  tablespace:3143595 role:3143588 diskquota exceeded  (seg1 127.0.0.1:6003 pid=3260)
-- clean up
DROP TABLE IF EXISTS t_in_custom_tablespace;
NOTICE:  table "t_in_custom_tablespace" does not exist, skipping
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u > /dev/null
SELECT diskquota.pause();
 pause 
-------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

DROP EXTENSION IF EXISTS diskquota;
\c contrib_regression;
DROP DATABASE IF EXISTS db_with_tablespace;
DROP TABLESPACE IF EXISTS custom_tablespace;
RESET ROLE;
DROP ROLE IF EXISTS role1;
DROP ROLE IF EXISTS role2;
